### PR TITLE
S3 object version id is not correctly updated during diff

### LINKED ISSUE #14899: Error/Bug: aws_s3_bucket_object inconsistent final plan with version_id
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v0.13.1
+ provider registry.terraform.io/-/aws v3.4.0
+ provider registry.terraform.io/hashicorp/aws v3.4.0
```

### Affected Resource(s)

* aws_s3_bucket_object

### Terraform Configuration Files

```hcl
provider "aws" {
  version                     = ">= 3.0.0"
}

resource "aws_s3_bucket" "examplebucket" {
  bucket = "sampletestexamplebuckettftest"
  acl    = "private"

  versioning {
    enabled = true
  }
}

variable "content" {
  type = string
}

resource "aws_s3_bucket_object" "firstfile" {
  key        = "someobject"
  bucket     = aws_s3_bucket.examplebucket.id
  content    = var.content
}

resource "aws_s3_bucket_object" "secondfile" {
  key        = "anotherobject"
  bucket     = aws_s3_bucket.examplebucket.id
  content    = "${md5(var.content)} - ${aws_s3_bucket_object.firstfile.version_id}"
}
```

### Debug Output

https://gist.github.com/wernerb/ec3872b69b1093a136fc9dc41880fb28

### Expected Behavior

Terraform knows that a new version_id is there and uses that correctly in `secondfile`

I have submitted a pull request that completely closes this issue.

### Actual Behavior

Error thrown during apply:

```
$ terraform apply -var content=bla2
aws_s3_bucket.examplebucket: Refreshing state... [id=sampletestexamplebuckettftest]
aws_s3_bucket_object.firstfile: Refreshing state... [id=someobject]
aws_s3_bucket_object.secondfile: Refreshing state... [id=anotherobject]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_s3_bucket_object.firstfile will be updated in-place
  ~ resource "aws_s3_bucket_object" "firstfile" {
        acl           = "private"
        bucket        = "sampletestexamplebuckettftest"
      ~ content       = "bla" -> "bla2"
        content_type  = "binary/octet-stream"
        etag          = "128ecf542a35ac5270a87dc740918404"
        force_destroy = false
        id            = "someobject"
        key           = "someobject"
        metadata      = {}
        storage_class = "STANDARD"
        tags          = {}
        version_id    = "5_Fxmip9w4DJgZV2or3_SAAn8Rr2QyA5"
    }

  # aws_s3_bucket_object.secondfile will be updated in-place
  ~ resource "aws_s3_bucket_object" "secondfile" {
        acl           = "private"
        bucket        = "sampletestexamplebuckettftest"
      ~ content       = "128ecf542a35ac5270a87dc740918404 - 5_Fxmip9w4DJgZV2or3_SAAn8Rr2QyA5" -> "a1288b1c7e2257a90bad9bdfb7690fbb - 5_Fxmip9w4DJgZV2or3_SAAn8Rr2QyA5"
        content_type  = "binary/octet-stream"
        etag          = "6be52a6a547f72e84fa7cdd36e01c56d"
        force_destroy = false
        id            = "anotherobject"
        key           = "anotherobject"
        metadata      = {}
        storage_class = "STANDARD"
        tags          = {}
        version_id    = "u04WlTsNLJ16GDUVqSxJ5Y0zAwGn7LT8"
    }

Plan: 0 to add, 2 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_s3_bucket_object.firstfile: Modifying... [id=someobject]
aws_s3_bucket_object.firstfile: Modifications complete after 0s [id=someobject]

Error: Provider produced inconsistent final plan

When expanding the plan for aws_s3_bucket_object.secondfile to include new
values learned so far during apply, provider
"registry.terraform.io/hashicorp/aws" produced an invalid new value for
.content: was cty.StringVal("a1288b1c7e2257a90bad9bdfb7690fbb -
5_Fxmip9w4DJgZV2or3_SAAn8Rr2QyA5"), but now
cty.StringVal("a1288b1c7e2257a90bad9bdfb7690fbb -
0mNGmhdBL8YmClTknce4vf3eQv5yFfF6").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Steps to Reproduce

```
1.terraform apply -var content=bla
```
First apply is ok.

Second setup
```
2. terraform apply -var content=anotherbla
```
Terraform applies correctly with the updated version_id.

### Important Factoids

Changes to `content` or similar keys that always produce changes to version should be hinted to terraform that it will be a new value. When using a version_id in a following resource when changing the object we get a inconsistent state error message from Terraform regarding the version_id.

Expected behavior is for terraform to be aware a `content` change is always a `version_id` change.

Please see connected merge request that fixes this issue.

### PR DESCRIPTION
**Pull request by wernerb**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

The configuration keys listed always result in changes to `version_id`. We make terraform aware of this. Currently terraform bails out during apply and refers to the AWS provider to fix this.
This also fixes unrelated bugs in the tests

Closes #14899

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_s3_bucket_object: version_id is correctly updated when certain configuration keys are changed
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSS3BucketObject'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSS3BucketObject -timeout 120m
=== RUN   TestAccAWSS3BucketObject_noNameNoKey
=== PAUSE TestAccAWSS3BucketObject_noNameNoKey
=== RUN   TestAccAWSS3BucketObject_empty
=== PAUSE TestAccAWSS3BucketObject_empty
=== RUN   TestAccAWSS3BucketObject_source
=== PAUSE TestAccAWSS3BucketObject_source
=== RUN   TestAccAWSS3BucketObject_content
=== PAUSE TestAccAWSS3BucketObject_content
=== RUN   TestAccAWSS3BucketObject_etagEncryption
=== PAUSE TestAccAWSS3BucketObject_etagEncryption
=== RUN   TestAccAWSS3BucketObject_contentBase64
=== PAUSE TestAccAWSS3BucketObject_contentBase64
=== RUN   TestAccAWSS3BucketObject_withContentCharacteristics
=== PAUSE TestAccAWSS3BucketObject_withContentCharacteristics
=== RUN   TestAccAWSS3BucketObject_NonVersioned
=== PAUSE TestAccAWSS3BucketObject_NonVersioned
=== RUN   TestAccAWSS3BucketObject_updates
=== PAUSE TestAccAWSS3BucketObject_updates
=== RUN   TestAccAWSS3BucketObject_updateSameFile
=== PAUSE TestAccAWSS3BucketObject_updateSameFile
=== RUN   TestAccAWSS3BucketObject_updatesWithVersioning
=== PAUSE TestAccAWSS3BucketObject_updatesWithVersioning
=== RUN   TestAccAWSS3BucketObject_updatesWithVersioningViaAccessPoint
=== PAUSE TestAccAWSS3BucketObject_updatesWithVersioningViaAccessPoint
=== RUN   TestAccAWSS3BucketObject_kms
=== PAUSE TestAccAWSS3BucketObject_kms
=== RUN   TestAccAWSS3BucketObject_sse
=== PAUSE TestAccAWSS3BucketObject_sse
=== RUN   TestAccAWSS3BucketObject_acl
=== PAUSE TestAccAWSS3BucketObject_acl
=== RUN   TestAccAWSS3BucketObject_metadata
=== PAUSE TestAccAWSS3BucketObject_metadata
=== RUN   TestAccAWSS3BucketObject_storageClass
=== PAUSE TestAccAWSS3BucketObject_storageClass
=== RUN   TestAccAWSS3BucketObject_tags
=== PAUSE TestAccAWSS3BucketObject_tags
=== RUN   TestAccAWSS3BucketObject_tagsLeadingSlash
=== PAUSE TestAccAWSS3BucketObject_tagsLeadingSlash
=== RUN   TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithNone
=== PAUSE TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithNone
=== RUN   TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithOn
=== PAUSE TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithOn
=== RUN   TestAccAWSS3BucketObject_ObjectLockRetentionStartWithNone
=== PAUSE TestAccAWSS3BucketObject_ObjectLockRetentionStartWithNone
=== RUN   TestAccAWSS3BucketObject_ObjectLockRetentionStartWithSet
=== PAUSE TestAccAWSS3BucketObject_ObjectLockRetentionStartWithSet
=== CONT  TestAccAWSS3BucketObject_noNameNoKey
=== CONT  TestAccAWSS3BucketObject_kms
=== CONT  TestAccAWSS3BucketObject_storageClass
=== CONT  TestAccAWSS3BucketObject_withContentCharacteristics
=== CONT  TestAccAWSS3BucketObject_updatesWithVersioningViaAccessPoint
=== CONT  TestAccAWSS3BucketObject_ObjectLockRetentionStartWithSet
=== CONT  TestAccAWSS3BucketObject_updatesWithVersioning
=== CONT  TestAccAWSS3BucketObject_updateSameFile
=== CONT  TestAccAWSS3BucketObject_ObjectLockRetentionStartWithNone
=== CONT  TestAccAWSS3BucketObject_updates
=== CONT  TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithOn
=== CONT  TestAccAWSS3BucketObject_NonVersioned
=== CONT  TestAccAWSS3BucketObject_acl
=== CONT  TestAccAWSS3BucketObject_metadata
=== CONT  TestAccAWSS3BucketObject_content
=== CONT  TestAccAWSS3BucketObject_tags
=== CONT  TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithNone
=== CONT  TestAccAWSS3BucketObject_contentBase64
=== CONT  TestAccAWSS3BucketObject_etagEncryption
=== CONT  TestAccAWSS3BucketObject_tagsLeadingSlash
--- PASS: TestAccAWSS3BucketObject_noNameNoKey (5.69s)
=== CONT  TestAccAWSS3BucketObject_sse
=== CONT  TestAccAWSS3BucketObject_updateSameFile
    resource_aws_s3_bucket_object_test.go:350: [INFO] Got non-empty plan, as expected
    resource_aws_s3_bucket_object_test.go:350: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSS3BucketObject_NonVersioned (48.64s)
=== CONT  TestAccAWSS3BucketObject_empty
--- PASS: TestAccAWSS3BucketObject_content (50.02s)
=== CONT  TestAccAWSS3BucketObject_source
--- PASS: TestAccAWSS3BucketObject_contentBase64 (50.38s)
--- PASS: TestAccAWSS3BucketObject_withContentCharacteristics (50.43s)
--- PASS: TestAccAWSS3BucketObject_etagEncryption (50.82s)
--- PASS: TestAccAWSS3BucketObject_kms (53.53s)
--- PASS: TestAccAWSS3BucketObject_sse (49.89s)
--- PASS: TestAccAWSS3BucketObject_updateSameFile (96.28s)
--- PASS: TestAccAWSS3BucketObject_empty (48.69s)
--- PASS: TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithOn (97.77s)
--- PASS: TestAccAWSS3BucketObject_updates (98.41s)
--- PASS: TestAccAWSS3BucketObject_source (49.26s)
--- PASS: TestAccAWSS3BucketObject_updatesWithVersioning (99.45s)
--- PASS: TestAccAWSS3BucketObject_updatesWithVersioningViaAccessPoint (107.91s)
--- PASS: TestAccAWSS3BucketObject_metadata (138.22s)
--- PASS: TestAccAWSS3BucketObject_ObjectLockRetentionStartWithNone (142.66s)
--- PASS: TestAccAWSS3BucketObject_acl (143.81s)
--- PASS: TestAccAWSS3BucketObject_ObjectLockLegalHoldStartWithNone (144.76s)
--- PASS: TestAccAWSS3BucketObject_ObjectLockRetentionStartWithSet (185.26s)
--- PASS: TestAccAWSS3BucketObject_tags (186.29s)
--- PASS: TestAccAWSS3BucketObject_tagsLeadingSlash (186.29s)
--- PASS: TestAccAWSS3BucketObject_storageClass (229.40s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	229.446
```

### CODE REVIEW

File: aws/resource_aws_s3_bucket_object.go [hunk 1]
```diff
@@ -527,8 +527,27 @@ func validateMetadataIsLowerCase(v interface{}, k string) (ws []string, errors [
...
+		"server_side_encryption",
+		"source",
+		"storage_class",
+		"website_redirect",
+	} {
+		if d.HasChange(key) {
+			return d.SetNewComputed("version_id")
+		}
+
 	}
```
Comment (gdavison) [inline aws/resource_aws_s3_bucket_object.go:551]: We have the function call `d.HasChanges()` that can take multiple keys

```suggestion
	if d.HasChanges("cache_control", "content_base64", "content_disposition", "content_encoding", "content_language", "content_type", "content", "etag", "kms_key_id", "metadata", "server_side_encryption", "source", "storage_class", "website_redirect") {
		return d.SetNewComputed("version_id")
	}
```

File: aws/resource_aws_s3_bucket_object_test.go [hunk 2]
```diff
@@ -1233,8 +1234,8 @@ resource "aws_s3_bucket_object" "object" {
   bucket                 = aws_s3_bucket.object_bucket.bucket
   key                    = "test-key"
   server_side_encryption = "AES256"
-  source                 = %[1]q
-  etag                   = filemd5(%[1]q)
+  source                 = %[2]q
+  etag                   = filemd5(%[2]q)
```
Comment (gdavison) [inline aws/resource_aws_s3_bucket_object_test.go:1238]: 👍 It might be worthwhile to use `%[1]d` to `bucket` for consistency here and below

**Review by gdavison — CHANGES_REQUESTED** [review]
Thanks for the PR, @wernerb. It looks great!

One change to make, and one suggestion

File: aws/resource_aws_s3_bucket_object_test.go (hunk 2 above)
Comment (wernerb) [reply aws/resource_aws_s3_bucket_object_test.go:1238]: I have changed all to be more consistent and updated the test output

File: aws/resource_aws_s3_bucket_object.go (hunk 1 above)
Comment (wernerb) [reply aws/resource_aws_s3_bucket_object.go:551]: @gdavison unfortunately `d.HasChanges` is not in `d.ResourceDiff`. I did refactor another instance in this file where we can use `d.hasChanges`.

**Comment by wernerb** [comment]
@gdavison I have implemented your changes.

**Comment by wernerb** [comment]
@gdavison I understand this is not the proper etiquette, but our company is currently suffering from this bug with cloudformation stacks unable to update without this change. This update hurts as it happened after the terraform 12 upgrade and we cannot go back anymore. 

Could you provide perhaps, at least an ETA so we can decide to create custom providers ourselves?

**Comment by wernerb** [comment]
@gdavison do you have a chance to look at my commenst on your comments. I rebased everything and the above test failure seems to be a problem in the pipeline?

**Comment by gdavison** [comment]
Thanks for your patience, @wernerb, and thanks again for the PR. The test failure was due to a timeout being too short.

**Pull request merged**